src/js/common/schemaValidatedComponent.js

```diff
@@ -0,0 +1,310 @@
+        return gpii.schema.component.validateComponent(shadowRecord.that);
+    };
+
+    /**
+     *
+     * Validate a component's options based on its schema.  For "potentia ii" (future) versions of Infusion, a "shadow"
+     * record is validated as part of the component's workflow.  For "legacy" (pre potentia II) versions, the record
+     * itself is validated on component creation.  The distinction is handled by a context-aware core grade that mixes
+     * in a grade with the correct strategy to use.
+     *
+     * @param {Object} componentToValidate - The component to validate.
+     */
+    gpii.schema.component.validateComponent = function (componentToValidate) {
+        if (fluid.componentHasGrade(componentToValidate, "gpii.schema.component")) {
+            var validationResults = gpii.schema.validator.validate(componentToValidate, componentToValidate.options.schema);
+            if (validationResults.isError) {
+                fluid.fail(validationResults.message);
+            }
+            else if (!validationResults.isValid) {
+                var localisedValidationErrors = gpii.schema.validator.localiseErrors(validationResults.errors);
+                var errorReport = "";
+                fluid.each(localisedValidationErrors, function (localisedError) {
+                    var failurePath = localisedError.dataPath.length ? localisedError.dataPath.join(" -> ") : "(root)";
+                    errorReport += "\n\t" + failurePath + ":\t" + localisedError.message;
+                });
+                fluid.fail("Component does not match its own schema, aborting component creation:" + errorReport);
+            }
+        }
+    };
+
+    gpii.schema.component.hasRegisterPotentia = function () {
+        return fluid.registerPotentia ? true : false;
+    };
+
+    fluid.contextAware.makeChecks({
+        "fluid.hasRegisterPotentia": "gpii.schema.component.hasRegisterPotentia"
+    });
+
+    // Validate the component as part of its startup "workflow".
+    fluid.defaults("gpii.schema.component.potentiaII", {
+        gradeNames: ["fluid.component"],
+        workflows: {
+            local: {
+                validateOptions: {
+                    priority: "after:concludeComponentObservation",
+                    funcName: "gpii.schema.component.validateShadowRecord",
+                    args:     ["{that}"]
+                }
+            }
+        }
+    });
+
+    // For components that don't have a workflow, validate on the `onCreate` event.
+    fluid.defaults("gpii.schema.component.legacy", {
+        gradeNames: ["fluid.component"],
+        listeners: {
+            "onCreate.validate": {
+                funcName: "gpii.schema.component.validateComponent",
+                args:     ["{that}"]
+            }
+        }
+    });
+
+    fluid.defaults("gpii.schema.component", {
+        gradeNames: ["fluid.component", "fluid.contextAware"],
+        schema: {
```

Comment:
Crikey, I really need to implement the "immutable options revolution". Just imagine a separate copy of this hanging around in memory for every instantiated component ........